src/internal/sdata/writers.go

```diff
@@ -0,0 +1,216 @@
+// package sdata deals with structured data.
+package sdata
+
+import (
+	"bufio"
+	"database/sql"
+	"encoding/csv"
+	"encoding/json"
+	"fmt"
+	"io"
+	"strconv"
+	"time"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+)
+
+// ErrTupleFields is returned by writers to indicate that
+// a tuple is not the right shape to be written to them.
+type ErrTupleFields struct {
+	Writer TupleWriter
+	Tuple  Tuple
+	Fields []string
+}
+
+func (e ErrTupleFields) Error() string {
+	return fmt.Sprintf("tuple has invalid fields for this writer (%T). expected: (%v), only have %d", e.Writer, e.Fields, (e.Tuple))
+}
+
+// TupleWriter is the type of Writers for structured data.
+type TupleWriter interface {
+	WriteTuple(row Tuple) error
+	Flush() error
+}
+
+type Tuple = []interface{}
+
+// CSVWriter writes Tuples in CSV format.
+type CSVWriter struct {
+	cw      *csv.Writer
+	headers []string
+
+	headersWritten bool
+	record         []string
+}
+
+// NewCSVWriter returns a CSVWriter writing to w.
+// If len(headers) == 0, then no headers will be written
+func NewCSVWriter(w io.Writer, headers []string) *CSVWriter {
+	return &CSVWriter{cw: csv.NewWriter(w)}
+}
+
+func (m *CSVWriter) WriteTuple(row Tuple) error {
```

Comment:
`len(row) == len(headers)`, always right?

Reply:
Yeah.  That should never come up, but if it did, it wouldn't panic with out of bounds below, you would just get more columns for the entry.

Reply:
makes sense! My only follow up question would be: would it make sense to include the validation present in JSONWriter below that throws `ErrTupleFields` error